pkg/controllers/networkpolicy/tracker.go

```diff
@@ -0,0 +1,283 @@
+	AppliedSecurityGroupDeleteError = "Detaching/Deleting security group %v: %v"
+)
+
+// CloudResourceNpTracker tracks NetworkPolicies applied on cloud resource.
+type CloudResourceNpTracker struct {
+	// CloudResource is a cloud resource.
+	CloudResource cloudresource.CloudResource
+	// NamespacedName of the cloud resource.
+	NamespacedName types.NamespacedName
+	// if dirty is true, cloud resource needs to recompute NetworkPolicy status.
+	dirty atomic.Value
+	// appliedToSGs is list of appliedToSecurityGroup to which cloud resource is a member.
+	appliedToSGs map[string]*appliedToSecurityGroup
+	// previously appliedToSGs to track sg clean up.
+	prevAppliedToSGs map[string]*appliedToSecurityGroup
+	// store appliedTo to network policies mapping.
+	appliedToToNpMap map[string][]types.NamespacedName
+	// NpStatus map of network policy (ANP) name to their realization status.
+	NpStatus map[string]*runtimev1alpha1.NetworkPolicyStatus
+}
+
+// newCloudResourceNpTracker create a tracker object using cloud resource object.
+// Currently, VM is the only cloud resource which has a tracker.
+func (r *NetworkPolicyReconciler) newCloudResourceNpTracker(rsc *cloudresource.CloudResource) *CloudResourceNpTracker {
+	log := r.Log.WithName("NPTracker")
+
+	vmItems, err := r.Inventory.GetVmFromIndexer(indexer.VirtualMachineByCloudResourceID, rsc.CloudResourceID.String())
+	if err != nil || len(vmItems) <= 0 {
+		log.Error(err, "failed to create np tracker for cloud resource id", "id", rsc.String())
+		return nil
+	}
+
+	// A vm is uniquely identified by its cloud assigned id and vpc id.
+	vm := vmItems[0].(*runtimev1alpha1.VirtualMachine)
+	tracker := &CloudResourceNpTracker{
+		appliedToSGs:     make(map[string]*appliedToSecurityGroup),
+		prevAppliedToSGs: make(map[string]*appliedToSecurityGroup),
+		appliedToToNpMap: make(map[string][]types.NamespacedName),
+		CloudResource:    *rsc,
+		// Use vm namespaced name for the tracker.
+		NamespacedName: types.NamespacedName{Name: vm.Name, Namespace: vm.Namespace},
+	}
+
+	log.Info("Create np tracker", "namespacedName", tracker.NamespacedName)
+	if err := r.cloudResourceNpTrackerIndexer.Add(tracker); err != nil {
+		log.Error(err, "add to CloudResourceNpTracker indexer")
+		return nil
+	}
+	return tracker
+}
+
+func (r *NetworkPolicyReconciler) getCloudResourceNpTracker(rsc *cloudresource.CloudResource, create bool) *CloudResourceNpTracker {
+	if obj, found, _ := r.cloudResourceNpTrackerIndexer.GetByKey(rsc.String()); found {
+		return obj.(*CloudResourceNpTracker)
+	} else if create {
+		return r.newCloudResourceNpTracker(rsc)
+	}
+	return nil
+}
+
+// processCloudResourceNpTrackers computes the np status of trackers marked dirty.
+func (r *NetworkPolicyReconciler) processCloudResourceNpTrackers() {
+	log := r.Log.WithName("NPTracker")
+	for _, obj := range r.cloudResourceNpTrackerIndexer.List() {
+		tracker := obj.(*CloudResourceNpTracker)
+		if tracker == nil || !tracker.isDirty() {
```

Comment:
Lets say we have 1000 VMs. We would be checking for 1000 VMs every sec for dirty flag. Can we optimize this to just iterate over dirty VMs? Please consider creating a temp dirty slice?